doc/man-source/lircd-uinput.8
.TH LIRCD-UINPUT "8" "Last change: Jun 2017" "lircd.uinput @version@" "System Managers Manual"

.\" Copyright (c) 2015, Alec Leamas
.\"
.\" %%%LICENSE_START(GPLv2+_DOC_FULL)
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, see
.\" <http://www.gnu.org/licenses/>.
.\" %%%LICENSE_END

.SH NAME
.P
lircd-uinput \- Forward lirc button presses as uinput events

.SH SYNOPSIS
.P
lircd-uinput [\fIoptions\fP]... [\fIlircd output socket\fP]
.br
lircd-uinput [\fIoptions\fP]... [\fIfile\fP]

.SH DESCRIPTION

.B lircd-uinput
reads decoded button press events from a
.BR lircd(8)
output socket and
injects these as key events to the kernel uinput device.
The kernel makes these events available to other applications on a
.I /dev/input/event
device, just as for regular input hardware.
.P
.B lircd-uinput
is designed to run as a systemd service.
.P
The
.I lircd output socket
argument is an optional path to the lircd output socket.
It defaults to the \fIoutput\fP value in the [lircd] section of the
lirc_options config file, falling back to
.I @varrundir@/lirc/lircd.
.P
If a plain file is given instead of a socket it is supposed to contain
button press events formatted as documented in
.BR lircd(8)
\&.
.B lircd-uinput
will in this case send the codes in this file and exit.
.P
In order to work, release events must be sent to the uinput device.
See REPEAT HANDLING.
.P
.B lircd-uinput
can only handle button events named using the official namespace i. e.,
buttons known to the kernel. See NOTES.
.P
Unless the
.I /dev/uinput
default permissions are changed
.B lirc-uinput
needs to run as root. See NOTES.
.P
The
.I /dev/input/event
device created by the kernel can be tricky to locate, and is not stable.
See NOTES.

.SH OPTIONS
.TP
\fB\-u\fR \fB\-\-uinput\fR <\fIdevice\fR>
uinput kernel device, defaults to /dev/uinput.
.TP
\fB\-r\fR \fB\-\-release-suffix\fR <\fIsuffix\fR>
.B lircd-uinput
generates different events depending on if the button name
ends with the \fIrelease-suffix\fR or not.
Defaults to
.I _EVUP
\&. See REPEAT HANDLING
.TP
\fB\-R\fR \fB\-\-repeat\fR <\fIdelay\fR[,\fIperiod\fR] | \fI,period\fR>
Set kernel autorepeat parameters of the created
.I /dev/input/event
device.
.I delay
is the time before repeating a key event,
.I period
is the time between repeated key events.
Both values are in milliseconds, separated by a comma.
Any  value can be omitted or set to 0 in which case
the kernel/system defaults are used.
See REPEAT HANDLING.
.TP
\fB\-a\fR \fB\-\-add-release-events\fR
Generate synthetic release events.
This option is mandatory unless
.BR lircd
is using the deprecated
.I --release
option.
The timeout used can be modified using the
.I release-timeout
key in the
.I lircd-uinput
section of the
.BR lirc_options.conf
file.
Useless if used with a text file as input.
\&. See REPEAT HANDLING
.TP
\fB\-d\fR \fB\-\-disabled\fR <\fIdisabled buttons file path\fR>
The path of a file which contains name of buttons to be disabled,
one per line.
By default,
.B lircd-uinput
sends all buttons found in input.
See NOTES.
.TP
\fB\-O\fR \fB\-\-options-file\fR <\fIpath\fR>
Options file holding default values. Defaults to
.I @etcdir@/lircd/lirc_options.conf
.TP
\fB-L, --logfile\fR <\fIlogfile path\fR>
Select the log file, either the string 'syslog' indicating that syslog(1)
should be used or a log file path.
The default is to use syslog.
.TP
\fB-D, --loglevel\fR [\fIlevel\fR]
Determine the amount of logging information.
.I level
can be a symbolic syslog level: 'error','warning, 'info', 'notice' or
\'debug'.
.B lircd-uinput
also defines three additional levels 'trace', 'trace1' and 'trace2' which
gives even more messages ('trace2' bringing the most).
However, in the log these messages are marked as 'debug'.
.TP
\fB\-h\fR \fB\-\-help\fR
Display usage summary.
.TP
\fB\-v\fR \fB\-\-version\fR
Display version info.

.SH REPEAT HANDLING

After receiving a keypress event the kernel will start emitting repeat
events to applications until
.B lircd-uinput
emits a corresponding release event.
The release events could be created by
.BR lircd(8)
using the deprecated
.I --release
option.
Such events typically has an
.I _EVUP
suffix appended to the original keypress event.
.P
The preferred way is that
.B lircd-uinput
creates release events using the
.I --add-release-events
option.
Combining both of these options might give unexpected results.
.P
Note that the linux kernel uses the suffix _UP (which was used py
lircd prior to 0.10.0) for other purposes since 4.7.
.P
In any case, the repeat events generated by the kernel can be tweaked using
the
.I --repeat
option. This can set the time between the keypress event and the
first repeat event and the time between each repeat event.
The
.I --repeat
option should only be used if no other program or udev rule is
automatically setting up key repeat parameters.
By design the repeat parameters can't be set in an atomic way
during input device registration so this option is prone to
race conditions.
.P
Note that the
.B lircd-uinput
repeat-event handling differs from
.BR lircd(8)
which  implements (emulates) hardware autorepeat in the events
sent to uinput (event.value == 2).
.B lircd-uinput
only sends keypress and release events, trusting the kernel autorepeat
support.

.SH NOTES

By default, the
.I dev/uinput
device is not writable for regular users.
Thus
.B lircd-uinput
needs to run as root unless these permissions are changed.
.P
One way to achieve this is an udev rule.
The lirc distribution contains an example rule which makes the
.B lircd-uinput
device writable for users in the group lirc.
Another way is to use
.BR lircd-setup(8)
to change the permissions. As distributed, the
.I lirc_options.conf
file contains an commented example using
.BR setfacl(1)
\&.

.B lircd-uinput
can only forward lircd button events with names from
the legal namespace defined by the kernel.
A complete list of possible button names  is available in
.I /usr/include/linux/input-event-codes.h
and also using
.I irrecord -l
\&.
Other button names are silently dropped.
.P
The event keycode depends on the name that was given a button in the
lircd config file e.g., if the button is named \fIKEY_1\fR the keycode
\fI1\fR will be generated.
.P
The kernel makes the events generated by
.B lircd-uinput
available on a
.I /dev/input
device like
.I /dev/input/event12.
This device name is not stable and typically varies after a reboot.
The lirc distribution contains a udev rule which if installed creates
a device link named
.I /dev/lircd-uinput
which always can be used.
.P
.B lircd-uinput
is designed to replace the currently deprecated
.I --uinput
option to
.BR lircd(8).
However,
.B lircd-uinput
forwards all events unless blocked by the
.I --disabled
option.
On the other hand,
.BR lircd
.I --uinput
does not forward all events - the heuristics used has been changed
over versions.

.SH BUGS

Since the device name is fixed, only one instance of
.B lircd-uinput
can run on a system.

.SH "SEE ALSO"

.br
.BR lircd (8)
.br
.BR irrecord(1)
.br
.BR http://www.kernel.org/doc/Documentation/input/event-codes.txt


NEWS
0.10.0-devel
 * The deprecated lircd --uinput option is dropped. Use
0.10.0-rc2 13/6/17
 * Fixed lirc-setup fail-on-start (#290).
 * Make python unit tests work on Debian (#289).
 * Fix mode2 config file parsing [292].
 * Documentation update.
 * Build fixes.
0.10.0rc1 28/5/17
 * ftdi driver: Fix off-by-2 timings (#275). This is a breaking
   change, see ftdi driver docs or source at doc/plugins/ftdi.html.
 * doc: The lirc.hwdb parsable driver list is deprecated - use
   the python database or the configs/*.yaml files instead.
 * lircd: the deprecated --uinput option is dropped. Use
   the lircd-uinput(8) daemon instead.
 * lircd: The --release option is deprecated.
 * lircd: Adding a --immediate-init option which initiates
   the hardware direct on start and exits on errors.
 * The slinke (a. k. a. slink-e) driver is dropped.
 * plugins: Add support for irdroid in the irtoy driver.
 * Enhanced device enumeration, mode2 --list-devices works for
   most drivers.
 * New, shiny (i. e., unstable) python API. See the API docs,
   module "python bindings". The API is also available as a pypi
   package, typically at /usr/share/lirc/python-pkg
 * lircd-uinput: Multiple config file parsing errors (#284).
 * The configs/*conf files no more supports  enumerating devices
   using shell scripts.
 * Added enumeration support and server version level check to
   driver API.
 * irrecord: Add --driver-options for lircd symmetry (#277).
 * lircd: Fix devices not connected at first attempt (#279).
 * lib: Fix failure when parsing absolute paths config files (#281).
 * lircd/irsend: Fix bug when sending/receiving concurrently for
   some remotes (#278, 5 pages!)
 * Added manual devinput configuration, cross-compile support (#264).
 * Various cross-compilation support patches (#264, #278)
 * ftdi driver: Fix off-by-2 timings (#275).
 * devinput driver: Fix bad codelength (#263).
 * OSX patches(#273)
 * Fix missing systemd support due to upstream changes (#274).
 * lircd, lircd-uinput: Change default release suffix due to linux
   kernel changes (#263).
 * Build: More quiet, enhanced V=0/V=1 support.

0.9.4d 23/01/17
 * The debian packaging means that freebsd and hurd builds are
   regularly tested, improving build stability.
 * A multitude of build patches, mostly after building on Debian:
    + Cross-compilation FTBS fixed (#233)
    + Fix configuration of configure (#234)
    + Fix library specification in lirc.pc (#236)
    + use HAVE_UINPUT=1 to force building uinput code (#238).
    + lib: Only use local headers in lirc_private.h (#237).
    + Configure libudev (#242)
    + Various kfreebsd fixes
    + Include config.h in installed headers (#250).
    + Fix _XOPEN_SOURCE related FTBS on some platforms (#248)
 * lib: Fix some lircrcd-related bugs (#254)
 * systemd: Fix wrong path to irexec (#239)
 * systemd: Add missing lircd-setup.service to distribution (#243)
 * lircd: Fix --listen option parsing [#249]
 * lircd: Fix segfault in SET_INPUTLOG without arguments (#252)
 * lircd: Fix DRV_OPTION error message handling (#256).
 * ircat: Fix option parsing bug + empty conditions (#251)
 * xmode2: Fix handling of --driver option.

0.9.4c 22/10/16
 * New curl_poll files adds a new MIT license. Lirc now carries
   GPL, BSD and MIT licenses.
 * Several patches to make lirc work again on macos (#226, #227).
 * New tool irtext2udp generates udp plugin data (#197).
 * Fix FTBS errors on linux kernel >= 4.8.0 (#232).
 * plugins: girs: Add udp/receive support (#201).
 * mode2: Fix broken --raw option (#223).
 * lirc-make-devinput: Fix that two remotes has the same name (#222).
 * lircd: Warn for duplicate remote definitions (#222).
 * lircd: Retry temporary unavailable write sockets (#221).
 * contrib: Fix usb devices' acl permissions (rhbz #1364744).
 * lirc-setup: Fix crasher bugfix (Fedora abort trace).
 * contrib: disable udev /dev/uinput rule requiring lirc user.
 * plugins: audio_alsa: Fix byte truncating in 16-bit data (#218):
 * plugins: irtoy: improve error handling (#220).
 * Minor documentation fixes.

0.9.4b 9/8/16
  * lircd-uinput: A multitude of fixes (#213, #211, #177). Adds new options.
  * lirc_client:  Fix "Can't redirect irsend output" (#207).
  * lirc_client: read_string() not terminating buffer contents (#216)
  * Remove a single python2 dependency [4e3c97c].
  * irrecord: Fix under-sized buffer causing crash (#208).
  * systemd: Add hardening options comments to services (#204).
  * lircmd: Options, udev examples and manpage update.A
  * lib: Add missing Makefile.am dep (#210).
  * doc: Manpages and configuration guide fixes.
  * build: Don't define docdir - autoconf defines docdir (#215).
  * build: support runstatedir configure option (#212).

0.9.4a 22/6/16
  * Fix syntax errors in systemd unit files (rhbz: 1350750).
  * irrecord: Enhance "Don't use devinput" message (#180).
  * plugins/uirt2_raw: Reset device to UIR mode on deinit.
  * plugins/girs.c: Enable reception after sending.
  * lirc.org: Updates and bugfixes
  * lircd, liblirc_client: Fix freeaddrinfo handling (#195).
  * Don't build commandir unless we have usb.h (#191).
  * lirc_options.conf: debug => loglevel (#199).
  * lib: use proper linking method to avoid parallel build issue
  * tools: make_rel_symlink.py can also use python2
  * plugins/devinput: Make the list-devices support available.
  * configure.ac: Bugfix (BSD setting reported wrong).

0.9.4 22/5/16
  * Issues: https://sf.net/p/lirc/tickets/milestone/0.9.4
  * Digest of 0.9.4pre1, 0.9.4pre2

  * The caraca driver has been retired (still available in attic/).
  * The iguanair driver has been moved out of tree and is now
    maintained by Iguanaworks, Inc. See drivers/iguanair/README.
  * The irman driver has been moved out of tree. From 0.5.1 the
    libirman package contains the driver. See drivers/irman/README

daemons/lircd-uinput.cpp
/**
 * @file lircd-uinput.c
 * This file implements the uinput forwarding service.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <poll.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/un.h>

#include <string>
#include <set>
#include <unordered_map>

#include <linux/input.h>
#include "lirc/input_map.h"

#include "lirc_private.h"
#include "line_buffer.h"


static const logchannel_t logchannel = LOG_APP;

/** The name used to register the uinput device. */
static const char* const DEVNAME = "lircd-uinput";

static const char* const HELP =
	"\nUsage: lircd-uinput [options] [socket]\n\n"
	"Argument:\n"
	"\t socket: lircd output socket or test file [" LIRCD "]\n"
	"\nOptions:\n"
	"\t -u --uinput=uinput \t\tuinput device [/dev/uinput]\n"
	"\t -r --release-suffix=suffix \tRelease events suffix [_EVUP]\n"
	"\t -R --repeat=delay[,period]\tSet kernel repeat parameters [none]\n"
	"\t -a --add-release-events\tAdd synthetic release events [no]\n"
	"\t -d --disable=file\t\tDisable buttons listed in file\n"
        "\t -D[level] --loglevel[=level]\t"
			"'info', 'warning', 'notice', etc., or 3..10.\n"
	"\t -L --logfile=file\t\tLog file path (default: use syslog)'\n"
	"\t -O --options-file\t\tOptions file"
			" [" SYSCONFDIR "/lirc/lirc_options.conf]\n"
	"\t -h --help\t\t\tDisplay this message\n"
	"\t -v --version\t\t\tDisplay version\n";


static const struct option cli_options[] = {
	{ "help",	        no_argument,       NULL, 'h' },
	{ "version",	        no_argument,       NULL, 'v' },
	{ "options-file",       required_argument, NULL, 'O' },
	{ "uinput",	        required_argument, NULL, 'u' },
	{ "release-suffix",     required_argument, NULL, 'r' },
	{ "repeat",             required_argument, NULL, 'R' },
	{ "add-release-events", no_argument,       NULL, 'a' },
	{ "loglevel",	        optional_argument, NULL, 'D' },
	{ "disable",	        required_argument, NULL, 'd' },
	{ "logfile",	        required_argument, NULL, 'L' },
	{ 0,		        0,		   0,    0   }
};


/** Used in parse_options(), matches cli_options above. */
static const char* const optstring = "ad:D::hO:r:R:u:L:v";

/** Max for --repeat period and delay parts (ms). */
static const int MAX_INTERVAL = 20000;

/** Keys in lirc_options db and thus in lirc_options.conf config file. */
static const char* const DEBUG_OPT    =	"lircd:debug";
static const char* const LOGFILE_OPT  =	"lircd-uinput:logfile";
static const char* const UINPUT_OPT   =	"lircd-uinput:uinput";
static const char* const SUFFIX_OPT   =	"lircd-uinput:release-suffix";
static const char* const TIMEOUT_OPT  =	"lircd-uinput:release-timeout";
static const char* const RELEASE_OPT  =	"lircd-uinput:add-release-events";
static const char* const REPEAT_OPT   =	"lircd-uinput:repeat";
static const char* const INPUT_ARG    =	"lircd-uinput:output";
static const char* const DISABLED_OPT =	"lircd-uinput:disabled";

/** Runtime options decoded from command line, config file and defaults. */
struct options {
	const char* uinput_path;    /**< The --uinput option. */
	const char* input_path;     /**< The socket argument. */
	const char* release_suffix; /**< The --release-suffix option. */
	unsigned release_timeout;   /**< Hidden release-timeout option. */
	unsigned repeat_delay;      /**< delay part of --repeat option. */
	unsigned repeat_period;     /**< period part of --repeat option. */
	bool add_release_events;    /**< The --add-release-events option. */
	loglevel_t loglevel;        /**< The --loglevel option. */
	const char* logfile;        /**< The --logfile option. */
	const char* disabled_path;  /**< The --disable file path option. */
	std::set<std::string>
		disabled_buttons;   /**< Disabled button names. */
	int inputfd;		    /**< input open  file descriptor. */
	int uinputfd;               /**< output open uinput descriptor. */
};


/** Cached code lookups for button names. */
class CodeCache {

	private:
		struct Entry {
			linux_input_code code;
			bool is_release;  /**< Button is a release event. */
		};

		typedef std::unordered_map<std::string, struct Entry> code_map;
		code_map cache;

	public:
		void add(const char* button, linux_input_code c, bool release)
		{
			Entry* entry = new Entry();
			entry->code = c;
			entry->is_release = release;
			cache[button] = *entry;
		};

		bool lookup(const std::string button_name,
			    linux_input_code* code,
			    bool* is_release)
		{
			auto it = cache.find(button_name);
			if (it == cache.end())
				return false;
			*code = it->second.code;
			*is_release = it->second.is_release;
			return true;
		};

		CodeCache() { cache = code_map(); }
};


/** Cache for keycode lookups. */
static CodeCache code_cache = CodeCache();

/** Set by send_message(), used when sending release events. */
static std::string last_button_press = std::string("");


/** Setup defaults for the CLI options parsing. */
static void add_defaults(void)
{
	char level[4];

	snprintf(level, sizeof(level), "%d", lirc_log_defaultlevel());
	const char* const suffix = options_getstring("lircd:release");
	const char* const socket = options_getstring("lircd:output");
	const char* const timeout = options_getstring(TIMEOUT_OPT);

	const char* const defaults[] = {
		DEBUG_OPT,		level,
		LOGFILE_OPT,		"syslog",
		UINPUT_OPT,		"/dev/uinput",
		REPEAT_OPT,	 	(const char*) NULL,
		SUFFIX_OPT,		suffix ? suffix : "_EVUP",
		TIMEOUT_OPT,		timeout ? timeout : "200",
		INPUT_ARG,		socket ? socket : LIRCD,
		DISABLED_OPT,		(const char*)NULL,
		(const char*)NULL,	(const char*)NULL
	};
	options_add_defaults(defaults);
}


/** Parse options and store in lirc_options db, argument to options_load(). */
static void parse_options(int argc, char** const argv)
{
	int c;

	optind = 1;
	add_defaults();
	c = getopt_long(argc, argv, optstring, cli_options, NULL);
	while (c != -1) {
		switch (c) {
		case 'h':
			fputs(HELP, stdout);
			exit(EXIT_SUCCESS);
		case 'v':
			printf("lircd-uinput %s\n", VERSION);
			exit(EXIT_SUCCESS);
		case 'O':
			break;
		case 'r':
			options_set_opt(SUFFIX_OPT, optarg);
			break;
		case 'R':
			options_set_opt(REPEAT_OPT, optarg);
			break;
		case 'u':
			options_set_opt(UINPUT_OPT, optarg);
			break;
		case 'a':
			options_set_opt(RELEASE_OPT, "True");
			break;
		case 'd':
			options_set_opt(DISABLED_OPT, optarg);
			break;
		case 'D':
			options_set_opt(DEBUG_OPT, optarg);
			break;
		case 'L':
			options_set_opt(LOGFILE_OPT, optarg);
			break;
		default:
			fputs("Usage: lirc.uinput [options] [config-file]\n",
                              stderr);
			exit(EXIT_FAILURE);
		}
		c = getopt_long(argc, argv, optstring, cli_options, NULL);
	}
	if (optind == argc - 1) {
		options_set_opt(INPUT_ARG, argv[optind]);
	} else if (optind != argc) {
		fputs("lircd-uinput: too many arguments (max one)\n", stderr);
		exit(EXIT_FAILURE);
	}
}


/** Handle the --disabled option, initiates the set of disabled buttons. */
static void parse_disabled(const char* path, std::set<std::string>* buttons)
{
	FILE* f;
	char buff[128];

	f = fopen(path, "r");
	if (f == NULL) {
		fprintf(stderr, "Cannot open %s for reading", path);
		exit(EXIT_FAILURE);
	}
	*buttons = std::set<std::string>();
	while (fgets(buff, sizeof(buff), f) != NULL) {
		std::string* button = new std::string(buff);
		const size_t pos = button->rfind('\n');
		if (pos != std::string::npos)
			button->erase(pos, 1);
		for (auto& c: *button)
			c = toupper(c);
		buttons->insert(*button);
	}
	fclose(f);
}


/** Handle the --repeat option, sets repeat_period and repeat_delay. */
static bool parse_repeat(struct options* opts)
{
	const char* repeat_opt = options_getstring(REPEAT_OPT);
	char optbuff[64];
	char* opt = optbuff;
	long value;

	opts->repeat_delay = 0;
	opts->repeat_period = 0;
	if (repeat_opt == NULL)
		return true;

	// Parsing: [delay][,period]
	strncpy(optbuff, repeat_opt, sizeof(optbuff) - 1);
	if (isdigit(*opt)) {
		value = strtol(opt, &opt, 10);
		if (value > MAX_INTERVAL || value < 0 )
			return false;
		opts->repeat_delay = static_cast<unsigned>(value);
	}
	if (opt == NULL || *opt == '\0')
		return true;
	if (*opt != ',')
		return false;
	opt++;
	if (*opt == '\0')
		return true; // Lets accept omitted period.
	if (!isdigit(*opt))
		return false;
	value = strtol(opt, &opt, 10);
	if (value > MAX_INTERVAL || value < 0)
		return false;
	if (opt != NULL && *opt != '\0')
		return false;
	opts->repeat_period = static_cast<unsigned>(value);
	return true;
}


/** Log some info on effective options. */
static void log_options(const struct options* opts)
{
	log_info("Reading data from %s, writing to %s",
		 opts->input_path, opts->uinput_path);
	if (opts->add_release_events) {
		log_info("Adding release events after a %d ms timeout",
			 opts->release_timeout);
	}
	if (opts->disabled_path != NULL) {
		log_info("Disabling %d key(s)",
			 opts->disabled_buttons.size());
	}
	if (opts->repeat_delay != 0) {
		log_info("Setting kernel repeat delay to %d ms",
			 opts->repeat_delay);
	}
	if (opts->repeat_period != 0) {
		log_info("Setting kernel repeat period to %d ms",
			 opts->repeat_period);
	}
	log_info("Using \"%s\" as release suffix", opts->release_suffix);
}


/** Get all options from lirc_options db and store in *opts. */
static void options_new(struct options* opts)
{
	opts->input_path = options_getstring(INPUT_ARG);
	opts->uinput_path = options_getstring(UINPUT_OPT);
	opts->logfile = options_getstring(LOGFILE_OPT);
	opts->disabled_path = options_getstring(DISABLED_OPT);
	opts->release_timeout = options_getint(TIMEOUT_OPT);
	opts->release_suffix = options_getstring(SUFFIX_OPT);
	opts->add_release_events = options_getboolean(RELEASE_OPT);
	if( !parse_repeat(opts)) {
		fputs("Warning: Cannot parse --repeat option.\n", stderr);
		fputs("Warning: Using kernel defaults\n", stderr);
		opts->repeat_delay = opts->repeat_period = 0;
	}
	opts->disabled_path = options_getstring(DISABLED_OPT);
	if (opts->disabled_path != NULL)
		parse_disabled(opts->disabled_path, &opts->disabled_buttons);
	const char* opt = options_getstring(DEBUG_OPT);
	opts->loglevel = string2loglevel(opt);
	if (opts->loglevel == LIRC_BADLEVEL) {
		fprintf(stderr, "Bad configuration loglevel:%s\n", opt);
		fprintf(stderr, HELP);
		fprintf(stderr, "Falling back to 'info'\n");
		opts->loglevel = LIRC_INFO;
	}
	opts->inputfd = -1;
	opts->uinputfd = -1;
}


/**
 * Get linux keycode for button name.
 *
 * @param button_name     Name of button.
 * @param suffix          The release suffix from the --release-suffix option.
 * @param[out] is_release If non-NULL, reflects if button_name is a regular
 *                        button with the suffix appended.
 * @return keycode or KEY_RESERVED if no match.
 */
static linux_input_code get_keycode(const char* button_name,
				    const char* suffix,
				    bool* is_release = NULL)
{
	linux_input_code input_code;
	if (is_release != NULL)
		*is_release = false;

	if (get_input_code(button_name, &input_code) != -1)
		return input_code;

	std::string s = button_name;
	size_t pos = s.rfind(suffix);
	if (pos == std::string::npos)
		return KEY_RESERVED;
	s = s.substr(0, pos);
	if (get_input_code(s.c_str(), &input_code) == -1)
		return KEY_RESERVED;
	if (is_release != NULL)
		*is_release = true;
	return input_code;
}


/** Send a struct input_event to an uinput fd, return success. */
static bool write_event(int fd, unsigned type, linux_input_code code, int val)
{
	struct input_event event;

	memset(&event, 0, sizeof(event));
	event.type = type;
	event.code = code;
	event.value = val;
	return write(fd, &event, sizeof(event)) == sizeof(event);
}


/** Given a button, format and send struct input_events to /dev/uinput. */
static void send_message(const struct options* opts, const char* button)
{
	linux_input_code code;
	bool is_release;

	button = button == NULL ? "(null)" : button;
	if (!code_cache.lookup(button, &code, &is_release)) {
		log_trace("Cache miss for %s", button);
		code = get_keycode(button, opts->release_suffix, &is_release);
		code_cache.add(button, code, is_release);
	}
	if (code == KEY_RESERVED) {
		log_info("Dropping non-standard symbol %s", button);
		return;
	}
	// event.value: 0 => release, 1 => press, 2 => autorepeat (not used).
	const int value = is_release ? 0 : 1;
	log_debug("Sending %s as %d:%d", button, code, value);

	if (!write_event(opts->uinputfd, EV_KEY, code, value))
		log_perror_err("Writing regular event to uinput failed");
	if (!write_event(opts->uinputfd, EV_SYN, SYN_REPORT, 0))
		log_perror_err("Writing EV_SYN to uinput failed");
	// send_release_event() needs to know if an event is required
	if (opts->add_release_events && !is_release)
		last_button_press = std::string(button);
	else
		last_button_press = "";
}


/** Process a single line of input from the socket (or test file). */
static void process_line(const struct options* opts, std::string line)
{
	int r;
	char button[PACKET_SIZE + 1];
	char remote[PACKET_SIZE + 1];
	int reps;

	r = sscanf(line.c_str(), "%*x %x %s %s\n", &reps, button, remote);
	if (r != 3) {
		log_warn("Cannot parse line: %s", line.c_str());
		return;
	}
	if (opts->disabled_path && opts->disabled_buttons.count(button) == 1)
		log_debug("Skipping disabled key %s", button)
	else
		send_message(opts, button);
}


/** Process all available lines in line_buffer. */
static void process_lines(const struct options* opts, LineBuffer* line_buffer)
{
	std::string line;

	while (line_buffer->has_lines()) {
		line = line_buffer->get_next_line();
		log_trace("Input: %s", line.c_str());
		process_line(opts, line);
	}
}


/**
 * If required, send a release event after a read timeout, emulating
 * a release event sent by lircd.
 */
static void send_release_event(const struct options* opts)
{
	if (!opts->add_release_events)
		return;
	if (last_button_press == "")
		return;
	std::string button_name = last_button_press + opts->release_suffix;
	send_message(opts, button_name.c_str());
	last_button_press = "";
}


/** The main worker: handle data, timeouts and errors on input fd. */
static void lircd_uinput(const struct options* opts)
{
	int r;
        char buffer[PACKET_SIZE + 1];
	LineBuffer line_buffer;
	struct pollfd fds;
	int timeout = opts->add_release_events ? opts->release_timeout : -1;

	while (true) {
		fds.fd = opts->inputfd;
		fds.events = POLLIN;
		fds.revents = 0;
		r = curl_poll(&fds, 1, timeout);
		if (r == 0) {
			send_release_event(opts);
			continue;
		}
		if ((fds.revents & POLLERR) != 0 || r < 0 ) {
			log_notice("POLLERR or curl_poll() error, exiting.");
			exit(EXIT_FAILURE);
		}
		r = read(opts->inputfd, buffer, PACKET_SIZE);
		if (r > 0) {
			line_buffer.append(buffer, static_cast<size_t>(r));
			process_lines(opts, &line_buffer);
		} else if (r == -1) {
			log_perror_warn("lircd_uinput(): read() error");
		} else  {
			fds.revents |= POLLHUP;
		}
		if ((fds.revents & POLLHUP) != 0) {
			log_debug("POLLHUP or no data: exiting .");
			exit(0);
		}
	}
}


/** Register all keys as active besides those --disabled. */
static bool register_keys(const options* opts, int fd)
{
	linux_input_code code;
	auto disabled_keys = std::set<linux_input_code>();

	for (auto button: opts->disabled_buttons) {
		code = get_keycode(button.c_str(), opts->release_suffix);
		disabled_keys.insert(code);
		log_trace("Disabling %s (%d)", button.c_str(), code);
	}
	for (int key = KEY_RESERVED; key <= KEY_MAX; key++) {
		if (disabled_keys.count(key) == 1) {
			log_debug("Cowardly refusing to register %d", key);
			continue;
		}
		if (ioctl(fd, UI_SET_KEYBIT, key) != 0)
			return false;
	}
	return true;
}


/** Try to setup the uinput output device fd. Returns valid fd or -1.  */
static int setup_uinputfd(const options* opts)
{
	int fd;
	struct uinput_user_dev dev;
	bool ok;

	fd = open(opts->uinput_path, O_RDWR);
	if (fd == -1) {
		log_perror_err("Cannot open uinput device: %s",
			       opts->uinput_path);
		return -1;
	}
	memset(&dev, 0, sizeof(dev));
	strncpy(dev.name, DEVNAME, sizeof(dev.name) - 1);
	ok = write(fd, &dev, sizeof(dev)) == sizeof(dev)
		&& ioctl(fd, UI_SET_EVBIT, EV_KEY) == 0
		&& ioctl(fd, UI_SET_EVBIT, EV_REP) == 0
		&& register_keys(opts, fd)
		&& ioctl(fd, UI_DEV_CREATE) == 0;
	if (ok)
		return fd;
	log_perror_err("could not setup uinput");
	close(fd);
	return -1;
}


/** Open the input file, a lircd socket or a plain file. Return fd or -1. */
static int open_input(const char* path)
{
	struct sockaddr_un addr;
	struct stat statbuf;
	int fd;

	if (stat(path, &statbuf) == -1) {
		log_perror_err("Cannot stat socket path %s", path);
		return -1;
	}
	if (S_ISREG(statbuf.st_mode)) {
		fd = open(path, O_RDONLY);
		if (fd >= 0)
			return fd;
		log_perror_err("Cannot open input file %s", path);
		return -1;
	}
	if (!S_ISSOCK(statbuf.st_mode)){
		log_perror_err("Unknown non-socket device %s", path);
		return -1;
	}
	addr.sun_family = AF_UNIX;
	strncpy(addr.sun_path, path, sizeof(addr.sun_path) - 1);
	fd = socket(AF_UNIX, SOCK_STREAM, 0);
	if (fd == -1) {
		log_perror_err("socket() failure");
		return -1;
	}
	if (connect(fd, (struct sockaddr*) &addr, sizeof(addr)) == -1) {
		log_perror_err("Cannot connect to socket %s", addr.sun_path);
		return -1;
	}
	return fd;
}


/** Set the event device REP_DELAY and REP_PERIOD parameters from *opts. */
static void set_kernel_repeat(const options* opts)
{
	unsigned delay = opts->repeat_delay;
	if (delay != 0) {
		log_debug("Setting kernel repeat delay to %d", delay);
		if (!write_event(opts->uinputfd, EV_REP, REP_DELAY, delay))
			log_warn("Cannot set kernel repeat delay");
	}
	unsigned period = opts->repeat_period;
	if (period != 0) {
		log_debug("Setting kernel repeat period to %d", period);
		if (!write_event(opts->uinputfd, EV_REP, REP_PERIOD, period))
			log_warn("Cannot set kernel repeat period");
	}
}


int main(int argc, char** argv)
{
	struct options opts = {0};

	options_load(argc, argv, NULL, parse_options);
	options_new(&opts);
	lirc_log_set_file(opts.logfile);
	lirc_log_open("lircd-uinput", 1, opts.loglevel);
	log_options(&opts);

	opts.inputfd = open_input(opts.input_path);
        if (opts.inputfd == -1) {
		log_error("Cannot setup input file descriptor.")
		exit(1);
	}
        opts.uinputfd = setup_uinputfd(&opts);
        if (opts.uinputfd == -1) {
		log_error("Cannot setup uinput output file descriptor.")
		exit(1);
	}
	set_kernel_repeat(&opts);
	lircd_uinput(&opts);
}
